File: frontend/src/gameLogic/Models/Game.ts
import TerritoryCardManager from '../Controllers/TerritoryCardManager';
import { Player } from './Player';

export class Game {
    private players: Array<Player>;
    private maxTurns: number;
    private currentTurn: number;
    private currentPlayersTurn: number;
    private hasPlayerCapturedAreaThisTurn: boolean;
    constructor(players: Array<Player>, maxTurns: number) {
        this.players = players;
        this.maxTurns = maxTurns;
        this.currentTurn = 0;
        this.currentPlayersTurn = 0;
        this.hasPlayerCapturedAreaThisTurn = false;
    }

    public getPlayers(): Array<Player> {
        return this.players;
    }

    public getCurrentPlayer(): Player {
        const indexOfPlayer = this.currentPlayersTurn;
        return this.players[indexOfPlayer];
    }

    public getTurnsRemaining(): number {
        return this.maxTurns - this.currentTurn;
    }

    public handlePlayerCapturingArea(): void {
        this.hasPlayerCapturedAreaThisTurn = true;
    }

    public handleNewTurn() {
        if (this.hasPlayerCapturedAreaThisTurn) {
            this.giveCurrentPlayerTerritoryCard();
        }

        this.changeCurrentPlayer();
        const newCurrentPlayer = this.getCurrentPlayer();
        newCurrentPlayer.addReinforcementsForNewTurn();
    }

    public changeCurrentPlayer() {
        this.currentPlayersTurn += 1;
        const lastPlayer = this.players.length - 1;
        if (this.currentPlayersTurn > lastPlayer) {
            this.currentPlayersTurn = 0;
            this.incrementCurrentTurn();
        }
    }

    public assignStartingUnits() {
        const unitsAvailable = this.getStartingUnitsAvailable();

        for (let i = 0; i < this.players.length; i++) {
            this.players[i].addReinforcements(unitsAvailable);
            this.players[i].addStartingUnits();
        }
    }

    private getStartingUnitsAvailable(): number {
        // TODO: remove after testing
        return 34;
        // if (this.players.length === 2) {
        //     return 60;
        // } else if (this.players.length === 3) {
        //     return 52;
        // } else {
        //     return 45;
        // }
    }

    public playersHaveReinforcements(): boolean {
        for (let i = 0; i < this.players.length; i++) {
            if (this.players[i].getReinforcements() > 0) {
                return true;
            }
        }

        return false;
    }

    private incrementCurrentTurn() {
        this.currentTurn++;
    }

    public checkMaxTurnsReached(): boolean {
        return this.maxTurns <= this.currentTurn;
    }

    public addUserIDToPlayer(userID: string) {
        for (let i = 0; i < this.players.length; i++) {
            if (this.players[i].getUserID() === userID) {
                return;
            } else if (!this.players[i].getUserID()) {
                this.players[i].setUserID(userID);
                return;
            }
        }
    }

    public waitingForUsersToJoin(): boolean {
        for (let i = 0; i < this.players.length; i++) {
            if (!this.players[i].getUserID()) {
                return true;
            }
        }

        return false;
    }

    private giveCurrentPlayerTerritoryCard(): void {
        const currentPlayer = this.getCurrentPlayer();
        TerritoryCardManager.givePlayerNewCard(currentPlayer);
        this.hasPlayerCapturedAreaThisTurn = false;
    }
}

File: server/game_logic/models/Game.py
import uuid
from game_logic.models.Player import Player
from game_logic.models.Area import Area
from typing import List, Dict
from game_logic.controllers.territory_cards import give_player_territory_card

class Game:
    def __init__(self, players: List[Player], areas: Dict[str, Area]):
        self.uuid = str(uuid.uuid4())[:8]
        self.players = players
        self.areas = areas
        self.current_turn = 0
        self.current_players_turn = 0
        self.has_player_captured_area_this_turn = False
        
    def handle_new_turn(self):
        if self.has_player_captured_area_this_turn:
            self.give_current_player_territory_card()
        
        self.has_player_captured_area_this_turn = False
        self.change_current_player()
        new_current_player = self.players[self.current_players_turn]
        new_current_player.add_reinforcements_for_new_turn()
    
    def give_current_player_territory_card(self):
        current_player = self.players[self.current_players_turn]
        give_player_territory_card(current_player)
    
    def change_current_player(self):
        self.current_players_turn += 1
        if self.current_players_turn > len(self.players) - 1:
            self.current_players_turn = 0
            self.current_turn += 1
            
    def handle_capturing_area(self, attacker: Player, area: Area):
        self.has_player_captured_area_this_turn = True 
        
        defender = area.player
        defender.remove_area(area)
        
        attacker.add_area(area)
        area.player = attacker
    
    def players_have_reinforcements(self) -> bool:
        for player in self.players:
            if player.reinforcements:
                return True 
        
        return False
        
    def turns_remaining(self) -> int:
        MAX_TURNS = 30
        return MAX_TURNS - self.current_turn
    
    def assign_starting_units(self):
        units_available = self.get_total_starting_units()
        for player in self.players:
            player.add_reinforcements(units_available)
            player.add_starting_units()
        
    def get_total_starting_units(self):
        # TODO: remove after testing
        return 34
        
        if len(self.players) == 2:
            return 60
        elif len(self.players) == 3:
            return 52
        else:
            return 45
        
    def add_user_id_to_next_avaiable_player(self, user_id: str):
        for player in self.players:
            if player.user_id == user_id:
                return 
            elif player.user_id == "":
                player.user_id = user_id
                return
            
    def num_players_left_to_join(self):
        total = 0
        
        for player in self.players:
            total += 1 if not player.user_id else 0
            
        return total